3.analysis/scripts/5.reconstruction.py

```diff
@@ -0,0 +1,195 @@
+dependency_df.head()
+
+
+# In[5]:
+
+
+# Initialize the MinMaxScaler
+scaler = MinMaxScaler()
+
+# Apply the scaler to the numeric columns
+dependency_df[dependency_df.select_dtypes(include=['float64', 'int']).columns] = scaler.fit_transform(
+    dependency_df.select_dtypes(include=['float64', 'int'])
+)
+
+
+# In[6]:
+
+
+train_and_test_subbed_dir = pathlib.Path("../0.data-download/data/train_and_test_subbed.parquet")
+train_and_test_subbed = pd.read_parquet(train_and_test_subbed_dir)
+
+train_and_test_subbed[train_and_test_subbed.select_dtypes(include=['float64', 'int']).columns] = scaler.fit_transform(
+    train_and_test_subbed.select_dtypes(include=['float64', 'int'])
+)
+
+# Convert DataFrame to NumPy and then Tensor
+train_test_array = train_and_test_subbed.to_numpy()
+train_test_tensor = torch.tensor(train_test_array, dtype=torch.float32)
+
+#Create TensorDataset and DataLoader
+tensor_dataset = TensorDataset(train_test_tensor)
+train_and_test_subbed_loader = DataLoader(tensor_dataset, batch_size=32, shuffle=False)
+
+
+# In[7]:
+
+
+results = []
+
+for model_file in model_save_dir.glob("*.joblib"):
+    model_file_name = model_file.stem
+    try:
+        parts = model_file_name.split("_")
+        model_name = parts[0]
+        num_components = int(parts[3])  # total latent dimensions
+        init = int(parts[7])  # initialization value
+        seed = int(parts[9])  # seed value
+    except (IndexError, ValueError):
+        print(f"Skipping file {model_file} due to unexpected filename format.")
+        continue
+
+    # Load the model
+    print(f"Loading model from {model_file}")
+    try:
+        model = joblib.load(model_file)
+    except Exception as e:
+        print(f"Failed to load model from {model_file}: {e}")
+        continue
+        
+    # Extract z, original input, and reconstruction
+    latent_df, original_data, reconstructed_data = extract_latent_dims(
+        model_name, model, dependency_df, train_and_test_subbed_loader, metadata
+    )
+    print(original_data)
+    print(reconstructed_data)
+
+    # Convert to torch tensors
+    original_tensor = torch.tensor(original_data, dtype=torch.float32)
+    reconstructed_tensor = torch.tensor(reconstructed_data, dtype=torch.float32)
+
+    # Clamp reconstructions to avoid log(0)
+    reconstructed_tensor = torch.clamp(reconstructed_tensor, min=1e-7, max=1 - 1e-7)
+
+    # Compute BCE loss across all elements
+    mse = F.mse_loss(reconstructed_tensor, original_tensor, reduction='mean')
+
+    results.append({
+        "model": model_name,
+        "latent_dim": num_components,
+        "init": init,
+        "mse": mse.item()
+    })
+    print("Original min/max:", original_data.min(), original_data.max())
+    print("Reconstructed min/max:", reconstructed_data.min(), reconstructed_data.max())
+
+# Convert results to DataFrame
+recon_df = pd.DataFrame(results)
+print(recon_df)
+
+
+# In[8]:
+
+
+import seaborn as sns
+import matplotlib.pyplot as plt
```

Comment:
Consider moving these imports to the start of the file. Python imports are almost always placed at the top of the file to help organize where this type of code appears.